src/bootstrap/bin/rustc.rs

```diff
@@ -104,6 +104,13 @@ fn main() {
         {
             cmd.arg("-C").arg("panic=abort");
         }
+
+        // Compile rustc crates with the more efficient initial-exec TLS model.
+        // This doesn't work with `dlopen`, so we can't use it by default in
+        // general, but we can use it for rustc's own internal libraries.
+        if crate_name.map_or(false, |n| n.starts_with("rustc_") || n.starts_with("rustc-")) {
+            cmd.arg("-Z").arg("tls-model=initial-exec");
+        }
```

Comment:
I wonder if we should also check that the path includes `compiler/rustc_` because there's some stuff on crates.io that also starts with `rustc_` or `rustc-` and it might not be strictly a `rustc_driver` dependency.

Reply:
We need a much better way to handle this, yes; this was just the simplest way to test the impact of this change.

Reply:
Best bet is probably to modify builder.rs to set RUSTC_IS_COMPILING_THE_COMPILER for cargo and then read that here.